diesel/src/mysql/connection/stmt/metadata.rs

```diff
@@ -1,52 +1,55 @@
-use std::collections::HashMap;
 use std::ffi::CStr;
+use std::ptr::NonNull;
 use std::slice;
 
 use super::ffi;
+use crate::mysql::connection::bind::Flags;
 
 pub struct StatementMetadata {
-    result: &'static mut ffi::MYSQL_RES,
-    column_indices: HashMap<&'static str, usize>,
+    result: NonNull<ffi::MYSQL_RES>,
 }
 
 impl StatementMetadata {
-    pub fn new(result: &'static mut ffi::MYSQL_RES) -> Self {
-        let mut res = StatementMetadata {
-            column_indices: HashMap::new(),
-            result,
-        };
-        res.populate_column_indices();
-        res
+    pub fn new(result: NonNull<ffi::MYSQL_RES>) -> Self {
+        StatementMetadata { result }
     }
 
-    pub fn fields(&self) -> &[ffi::MYSQL_FIELD] {
+    pub fn fields(&'_ self) -> &'_ [MysqlFieldMetadata<'_>] {
         unsafe {
-            let ptr = self.result as *const _ as *mut _;
-            let num_fields = ffi::mysql_num_fields(ptr);
-            let field_ptr = ffi::mysql_fetch_fields(ptr);
-            slice::from_raw_parts(field_ptr, num_fields as usize)
+            let num_fields = ffi::mysql_num_fields(self.result.as_ptr());
+            let field_ptr = ffi::mysql_fetch_fields(self.result.as_ptr());
+            if field_ptr.is_null() {
+                &[]
+            } else {
+                slice::from_raw_parts(field_ptr as _, num_fields as usize)
+            }
         }
     }
+}
 
-    pub fn column_indices(&self) -> &HashMap<&str, usize> {
-        &self.column_indices
+impl Drop for StatementMetadata {
+    fn drop(&mut self) {
+        unsafe { ffi::mysql_free_result(self.result.as_mut()) };
     }
+}
+
+#[repr(transparent)]
+pub struct MysqlFieldMetadata<'a>(ffi::MYSQL_FIELD, std::marker::PhantomData<&'a ()>);
 
-    fn populate_column_indices(&mut self) {
-        self.column_indices = self
-            .fields()
-            .iter()
-            .enumerate()
-            .map(|(i, field)| {
-                let c_name = unsafe { CStr::from_ptr(field.name) };
-                (c_name.to_str().unwrap_or_default(), i)
-            })
-            .collect()
+impl<'a> MysqlFieldMetadata<'a> {
+    pub fn field_name(&self) -> Option<&str> {
+        if self.0.name.is_null() {
+            None
+        } else {
+            unsafe { CStr::from_ptr(self.0.name).to_str().ok() }
```

Comment:
It's different in that way that we have some way to handle non `UTF-8` field names here, we can just return `None` in that case.

Reply:
I'm not sure I understand. On the other backends, we decided to not silent-fail on this and `expect`, because they should never return non utf-8.
On this backend it seems we decided to return them as non-named fields.
Is it because in the way we're using Mysql now it is expected behaviour that it might return non-utf-8 values?

Oh well nevermind I guess that's probably the correct solution if the previous correct solution was what was previously implemented, that is `c_name.to_str().unwrap_or_default()` ^^